cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -0,0 +1,1588 @@
+
+        :rtype None
+        """
+
+        # Checks that adata is fully loaded
+        if self.adata.isbacked:
+            raise RuntimeError(
+                "adata is loaded in backed mode, validation requires anndata to be loaded in memory"
+            )
+
+        # Checks that reserved columns are not used
+        self._check_column_availability()
+
+        # Checks sparsity
+        self._validate_sparsity()
+
+        # Checks each component
+        for component, component_def in self.schema_def["components"].items():
+
+            # Skip if component does not exist: only useful for adata.raw.var
+            if _getattr_anndata(self.adata, component) is None:
+                if "required" in component_def:
+                    self.errors.append(
+                        f"'{component}' is missing from adata and it's required"
+                    )
+                continue
+
+            if component_def["type"] == "dataframe":
+                self._validate_dataframe(component)
+            elif component_def["type"] == "dict":
+                dictionary = getattr(self.adata, component)
+                self._validate_dict(dictionary, component, component_def)
+            elif component_def["type"] == "embedding_dict":
+                self._validate_embedding_dict()
+            else:
+                raise ValueError(f"Unexpected component type '{component['type']}'")
+
+        # Checks for raw only if there are no errors, because it depends on the
+        # existence of adata.obs["assay_ontology_term_id"] and adata.obs["X_normalization"]
+        if not self.errors and "raw" in self.schema_def:
+            self._validate_raw()
+
+    def validate_adata(self, h5ad_path: Union[str, bytes, os.PathLike] = None) -> bool:
+
+        """
+        Validates adata
+
+        :params Union[str, bytes, os.PathLike] h5ad_path: path to h5ad to validate, if None it will try to validate
+        from self.adata
+
+        :return True if successful validation, False otherwise
+        :rtype bool
+        """
+
+        # Re-start errors in case a new h5ad is being validated
+        self.errors = []
+
+        if h5ad_path:
+            self._read_h5ad(h5ad_path)
+
+        # Fetches schema def from anndata if schema version is not found in AnnData, this fails
+        self._set_schema_def()
+
+        if not self.errors:
+            self._deep_check()
+
+        # Print warnings if any
+        if self.warnings:
+            self.warnings = ["WARNING: " + i for i in self.warnings]
+            print(*self.warnings, sep="\n")
+
+        # Print errors if any
+        if self.errors:
+            self.errors = ["ERROR: " + i for i in self.errors]
+            print(*self.errors, sep="\n")
+            self.is_valid = False
+        else:
+            self.is_valid = True
+
+        return self.is_valid
+
+
+class LabelWriter:
```

Comment:
Can you expand on why it's worth moving it? 

`H5adLabelAppender` is initialized with a`Validator` instance. And both classes are used in the function `validate` which is the entry point of the CLI.

They all are in the same file for convenience.

Reply:
Primarily to reduce the size of the file and improve readability.  As a reader of the code, scanning the files/packages is usually my first introduction to the structure of the code and the various components. (It's unfortunate IDE are still file-oriented, rather than class-oriented!).  Anyway, quite minor, no strong reason to change it.